Adds Metal Support to ociodisplay and enables running GPU tests with Metal backend

Morteeza opened the pull request.

hodoulp commented:
Thanks for the MSL support for unit tests, I put a first look to the work (and differ the in-depth review to next week) to at least see the changes. Below is an unordered list of questions/comments:
-   The first question is around the decision to add an extra cmd line argument for the gpu unit tests to run in metal instead of creating another unit test dedicated to MSL like discussed. Please change nothing that's only questions/comments to better understand.
      - I currently have no pros/cons arguments between the two approaches. But I would like to know if you faced some challenge(s) guiding or imposing this approach. 
      - On one side, the ogl library (which stands for OpenGL) originally only offered OpenGL utilities to help build OpenGL apps. Adding Metal support is at first glance adding some confusion. But the library could be enhanced to offer several GPU environment supports to ease unit tests and apps writings. _Need more time to think about!_
       - Adding an extra command line argument to the existing GPU unit test is an elegant way to use the same unit tests for all the GPU pipelines which is a good thing. _Need more time to think about!_
- `ociodisplay` the `-gpuinfo` (which dumps the complete GPU shader code to the console) does not work for MSL
- I noticed some `assert` and `Exception`. The rule is to always use the `OCIO::exception` to report errors.
- There are several GPU unit tests failing. Listed below are some of them in case you do not have the same:
    - `[ 27/207] [ExposureContrast / style_linear_dynamic_parameter ] `, it seems to do nothing
    - `[ 28/207] [ExposureContrast / dp_several_one_dynamic         ]`, error on all color components
    - `[ 29/207] [ExposureContrast / dp_several_both_dynamic        ]`, error on all color components
    - `[ 80/207] [GradingPrimary / style_log_dynamic_retests        ]`, `blue` is completely wrong
    - `[ 81/207] [GradingPrimary / two_transforms_retests           ]`, `red` is completely wrong
    - `[ 90/207] [GradingRGBCurve / style_log_dynamic_retests       ]`, `red` is completely wrong
    - etc.

hodoulp commented on include/OpenColorIO/OpenColorIO.h:
```diff
@@ -2825,10 +2825,12 @@ class OCIOEXPORT GpuShaderCreator
                             const Float3Getter & getFloat3) = 0;
 
     virtual bool addUniform(const char * name,
+                            const size_t & maxSize,
```
@Morteeza The public API cannot be changed like that because the MSL translation is planned for a patch version. 
So, that change is a blocker for OCIO 2.1.1 !

Morteeza replied:
@hodoulp got it. I will change this to not touch the public API.

Morteeza commented:
> Thanks for the MSL support for unit tests, I put a first look to the work (and delay the in-depth review to next week) to at least see the changes. Below is an unordered list of questions/comments:
> 
> * The first question is around the decision to add an extra cmd line argument for the gpu unit tests to run in metal instead of creating another unit test dedicated to MSL like discussed. Please change nothing that's only questions/comments to better understand.
>   
>   * I currently have no pros/cons arguments between the two approaches. But I would like to know if you faced some challenge(s) guiding or imposing this approach.
>   * On one side, the ogl library (which stands for OpenGL) originally only offered OpenGL utilities to help build OpenGL apps. Adding Metal support is at first glance adding some confusion. But the library could be enhanced to offer several GPU environment supports to ease unit tests and apps writings. _Need more time to think about!_
>   * Adding an extra command line argument to the existing GPU unit test is an elegant way to use the same unit tests for all the GPU pipelines which is a good thing. _Need more time to think about!_
> * `ociodisplay` the `-gpuinfo` (which dumps the complete GPU shader code to the console) does not work for MSL
> * I noticed some `assert` and `Exception`. The rule is to always use the `OCIO::exception` to report errors.
> * There are several GPU unit tests failing. Listed below are some of them in case you do not have the same:
>   
>   * `[ 27/207] [ExposureContrast / style_linear_dynamic_parameter ] `, it seems to do nothing
>   * `[ 28/207] [ExposureContrast / dp_several_one_dynamic         ]`, error on all color components
>   * `[ 29/207] [ExposureContrast / dp_several_both_dynamic        ]`, error on all color components
>   * `[ 80/207] [GradingPrimary / style_log_dynamic_retests        ]`, `blue` is completely wrong
>   * `[ 81/207] [GradingPrimary / two_transforms_retests           ]`, `red` is completely wrong
>   * `[ 90/207] [GradingRGBCurve / style_log_dynamic_retests       ]`, `red` is completely wrong
>   * etc.

As you mentioned having an extra argument and switch to metal tests, let us to reuse lots of code. 

But regarding the decision to not make an abstraction around rendering code and having OpenGL and Metal implementation, the main problem there is GLUT lib that is used for OpenGL application implementation. All UI work needs to be redone if we try to go that way. Therefore, I decided to have Metal rendering to a texture that is shared between GL and Metal, and then display that on the screen with GL. This way we don't need to touch UI code. Also, probably rest of OCIO rendering apps can be ported this way, too.

Btw, `-gpuinfo` should be fixed, now. Also I switched all asserts to `Exception`, and I fixed gpu rendering tests failings that you mentioned. I noticed depending on GPU vendor, there are difference in output results. I am investigating that and will address them, soon.

hodoulp commented:
Starting some tests before doing the in-depth code review, below is an unordered list of findings:

The commands are:
1. `export OCIO=/Users/hodoulp/dev/OpenColorIO-Configs/aces_1.0.3/config.ocio`
2. `lldb ./src/apps/ociodisplay/ociodisplay -- -v -gpuinfo -metal`

The findings are:
- using `ociodisplay` in debug mode and the [ACES 1.0.3](https://github.com/imageworks/OpenColorIO-Configs) config. The app successfully starts with some exceptions: 
```
2021-12-03 13:32:49.817127-0500 ociodisplay[13305:2504721] [General] -[__NSCFString replaceCharactersInRange:withString:]: Range or index out of bounds
2021-12-03 13:32:49.818170-0500 ociodisplay[13305:2504721] [General] (
	0   CoreFoundation                      0x00007fff375b4437 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff7058c5bf objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff37663e8e -[__NSCFString characterAtIndex:].cold.1 + 0
	3   CoreFoundation                      0x00007fff37664105 -[__NSCFString replaceOccurrencesOfString:withString:options:range:].cold.1 + 0
	4   CoreFoundation                      0x00007fff375f7178 mutateError + 44
	5   CoreFoundation                      0x00007fff3754acc8 -[__NSCFString replaceCharactersInRange:withString:] + 50
	6   GLUT                                0x00007fff39f76851 -[GLUTApplication applicationWillFinishLaunching:] + 175
	7   CoreFoundation                      0x00007fff3752e0af __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
	8   CoreFoundation                      0x00007fff3752e043 ___CFXRegistrationPost1_block_invoke + 63
	9   CoreFoundation                      0x00007fff3752dfb8 _CFXRegistrationPost1 + 372
	10  CoreFoundation                      0x00007fff3752dc24 ___CFXNotificationPost_block_invoke + 80
	11  CoreFoundation                      0x00007fff374fdd8d -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1554
	12  CoreFoundation                      0x00007fff374fd239 _CFXNotificationPost + 1351
	13  Foundation                          0x00007fff39b7d086 -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
	14  AppKit                              0x00007fff3479be99 -[NSApplication finishLaunching] + 330
	15  AppKit                              0x00007fff34a85911 _NSApplicationBeginRunning + 179
	16  AppKit                              0x00007fff34c3d125 -[NSApplication _beginRunning] + 26
	17  GLUT                                0x00007fff39f76523 -[GLUTApplication run] + 117
	18  GLUT                                0x00007fff39f827e7 glutMainLoop + 264
	19  ociodisplay                         0x0000000100007574 main + 2100
	20  libdyld.dylib                       0x00007fff71734cc9 start + 1
)
2021-12-03 13:32:49.818638-0500 ociodisplay[13305:2504721] [General] _createMenuRef called with existing principal MenuRef already associated with menu
2021-12-03 13:32:49.819250-0500 ociodisplay[13305:2504721] [General] (
	0   CoreFoundation                      0x00007fff375b4437 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff7058c5bf objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff375b4295 +[NSException raise:format:] + 189
	3   AppKit                              0x00007fff347a3612 -[NSCarbonMenuImpl _createMenuRef] + 55
	4   AppKit                              0x00007fff347a2fe2 -[NSCarbonMenuImpl _instantiateCarbonMenu] + 133
	5   AppKit                              0x00007fff3479c005 -[NSApplication finishLaunching] + 694
	6   AppKit                              0x00007fff34a85911 _NSApplicationBeginRunning + 179
	7   AppKit                              0x00007fff34c3d125 -[NSApplication _beginRunning] + 26
	8   GLUT                                0x00007fff39f76523 -[GLUTApplication run] + 117
	9   GLUT                                0x00007fff39f827e7 glutMainLoop + 264
	10  ociodisplay                         0x0000000100007574 main + 2100
	11  libdyld.dylib                       0x00007fff71734cc9 start + 1
)
```
- right click on the grey scale image to change the transform to `DCDM`:
```
Process 13305 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
    frame #0: 0x00007fff7057f81d libobjc.A.dylib`objc_msgSend + 29
libobjc.A.dylib`objc_msgSend:
->  0x7fff7057f81d <+29>: andl   0x18(%r10), %r11d
    0x7fff7057f821 <+33>: shlq   $0x4, %r11
    0x7fff7057f825 <+37>: addq   0x10(%r10), %r11
    0x7fff7057f829 <+41>: cmpq   (%r11), %rsi
Target 0: (ociodisplay) stopped.
(lldb) thread backtrace
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
  * frame #0: 0x00007fff7057f81d libobjc.A.dylib`objc_msgSend + 29
    frame #1: 0x00000001000226b1 ociodisplay`OpenColorIO_v2_2dev::MetalBuilder::applyColorCorrection(this=0x000000010913bc40, inputTexture=0x0000000119c21780, outputTexture=0x0000000119cf4570, outWidth=512, outHeight=512) at msl.mm:454:38
    frame #2: 0x0000000100030d07 ociodisplay`OpenColorIO_v2_2dev::MetalApp::redisplay(this=0x0000000109025378) at metalapp.mm:366:25
    frame #3: 0x00000001000052a2 ociodisplay`Redisplay() at main.cpp:209:19
    frame #4: 0x00007fff39f6db21 GLUT`-[GLUTView drawRect:] + 93
    frame #5: 0x00007fff39f6d88e GLUT`-[GLUTView handleWorkEvent:] + 3420
    frame #6: 0x00007fff39f82592 GLUT`processWindowWorkList + 67
    frame #7: 0x00007fff39f82530 GLUT`__glutProcessWorkEvents + 32
    frame #8: 0x00007fff39f7658d GLUT`-[GLUTApplication run] + 223
    frame #9: 0x00007fff39f827e7 GLUT`glutMainLoop + 264
    frame #10: 0x0000000100007574 ociodisplay`main(argc=4, argv=0x00007ffeefbff938) at main.cpp:754:5
    frame #11: 0x00007fff71734cc9 libdyld.dylib`start + 1
    frame #12: 0x00007fff71734cc9 libdyld.dylib`start + 1
(lldb) 
```

Last running `./tests/cpu/test_gpu_exce -metal` fully succeeds but it generates few identical messages:

- ```[ 98/207] [GradingTone / style_log_highlights_rev            ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 0 at pix[0][0])```
- ```[148/207] [Lut1DOp / lut1d_1_small_legacy_shader             ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 6.34193e-05 at pix[41424][2]```
- ```[198/207] [MatrixOps / matrix_offset_generic_shader          ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 4.76837e-07 at pix[46153][1])```

hodoulp commented on tests/gpu/GPUUnitTest.cpp:
```diff
@@ -319,10 +322,10 @@ namespace
     {
         app->setPrintShader(test->isVerbose());
-
+        
```
Remove these extra spaces.

hodoulp commented on tests/gpu/GPUUnitTest.cpp:
```diff
@@ -501,13 +504,37 @@ namespace
     }
 };
 
+#if __APPLE__
```
From line 507 to line 525, the code can loop on arguments for all platforms, but only macOS has an argument to catch

hodoulp reviewed (changes requested):
Lots of comments but great progress.

Comments to highlight are:
- Only add C++ code
- there is a macOS 10.15 limitation, why?
- adding array size to method argument list, why?
- GPU keyword `any` need ?
- etc.

hodoulp commented:
> Therefore, I decided to have Metal layer to render to a texture that is shared between GL and Metal, and then display that on the screen with GL. This way we don't need to touch UI code. Also, probably rest of OCIO rendering apps can be ported this way, too. As I mentioned, I tried to make this change as least intrusive as possible.

I now understand the rational. My last comments are:

- on macOS, the usual way to run unit tests i.e. `ctest`, `ctest -V`, etc. will never run the metal unit tests. _To think about._
- on macOS, `oglappHelpers` runs OpenGL with backend supporting GLSL & Metal is finally quite convincing.

hodoulp commented:
> on macOS, the usual way to run unit tests i.e. ctest, ctest -V, etc. will never run the metal unit tests. _To think about._

Add the following lines to []/tests/gpu/CMakeLists.txt at line 55:
```
if(APPLE)
	add_test(NAME test_metal COMMAND test_gpu_exec -metal)
endif()
```

Morteeza commented on src/OpenColorIO/GpuShaderClassWrapper.cpp:
```diff
@@ -27,7 +32,9 @@ std::string MetalShaderClassWrapper::generateClassWrapperHeader(GpuShaderText& k
     for(const auto& param : m_functionParameters)
     {
-        kw.newLine() << separator << param.type << " " << param.name;
+        kw.newLine() << separator << (param.isArray ? "constant " : "") << param.type << " " << param.name;
```
constant specifies address space of the array/pointer. It specifies that the value is read-only and act like a uniform variable in OpenGL. Without the address space, we can't pass constant buffer passed arrays to this function.

Morteeza commented on src/OpenColorIO/GpuShaderClassWrapper.cpp:
```diff
@@ -38,20 +45,28 @@ std::string MetalShaderClassWrapper::generateClassWrapperHeader(GpuShaderText& k
             
             kw.newLine()    << "for(int i = 0; i < "
-                            << param.name.substr(openAngledBracketPos+1, closeAngledBracketPos-openAngledBracketPos-1)
+                            << GetArrayLengthVariableName(variableName)
```
The array size specifies passed data size upper limit, but in practice what we pass as constant/uniform buffer contains less data. To generate entry function and constant buffer definition (in msl.mm), this upper limit needs to be known but information is not available and to make it available, I either have to parse the shader and extract max array size or add max length member to `VectorFloat` and `VectorInt`. First one is very dirty, and second one is clean but touches OpenColorIO.h. Therefore, I fixed the problem by passing how many elements of this array is passed through constant/uniform buffer. 

I actually like this approach because this way the size of data needs to be passed to GPU is reduced.

Morteeza commented on src/OpenColorIO/GpuShaderClassWrapper.h:
```diff
@@ -64,10 +64,13 @@ class MetalShaderClassWrapper : public GpuShaderClassWrapper
             type(type),
             name(name)
```
added `m_` but it seems to be compiled on all platforms.

Morteeza commented on src/libutils/oglapphelpers/metalapp.mm:
```diff
@@ -0,0 +1,380 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
You mean instead of objective-C? We can go for metalcpp bindings, but I feel objective-c is the better choice that gives us more control on what happens. After all, objective-c and swift are the main languages for implementing metal apps. Since this is apple platform specific code, that shouldn't be a problem.

hodoulp replied:
Good point.

Morteeza commented on src/libutils/oglapphelpers/oglapp.h:
```diff
@@ -130,7 +130,14 @@ class OglApp
 
+    bool printShader() const { return m_printShader; }
+    
+    OpenGLBuilderRcPtr m_oglBuilder;
```
that's already protected.

Morteeza commented on src/libutils/oglapphelpers/msl.mm:
```diff
@@ -0,0 +1,502 @@
+
+void MetalBuilder::triggerProgrammaticCaptureScope()
+{
+    if (@available(macOS 10.15, *)) {
```
we actually don't need `triggerProgrammaticCaptureScope` and `stopProgrammaticCaptureScope` to be part of msl.mm. They are only for debug purposes. We can remove them.

Morteeza replied:
I would rather keep these and add a message that triggering capture is only supported on MacOS 10.15 and above. Since this is not a core functionality but more for debugging, that should be ok.

Morteeza commented on src/libutils/oglapphelpers/mtltexture.h:
```diff
@@ -0,0 +1,75 @@
+    unsigned int                    m_texID;
+    id<MTLTexture>                  m_metalTexture;
+    
+    const GLMetalTextureFormatInfo* m_formatInfo;
```
that's a pointer to static array. mtltexture is not supposed to delete it.

Morteeza commented on src/libutils/oglapphelpers/mtltexture.h:
```diff
@@ -0,0 +1,75 @@
+    void createMetalTexture();
+    
+    id<MTLDevice>                   m_device;
+    NSOpenGLContext*                m_openGLContext;
```
good catch! Thanks... I set it to be in autorelease pool now.

Morteeza commented on tests/cpu/GpuShader_tests.cpp:
```diff
@@ -1138,21 +1138,33 @@ OCIO_ADD_TEST(GpuShader, MetalSupport9)
-  , int ocio_grading_rgbcurve_coefsOffsets[8]
-  , float ocio_grading_rgbcurve_coefs[180]
+  constant int ocio_grading_rgbcurve_knotsOffsets[8]
+  , int ocio_grading_rgbcurve_knotsOffsets_count
```
Because in current use cases, we don't fill all elements of this array with constant/uniform buffer data necessarily. Therefore, we only set it partially and set the rest to zero.

Morteeza commented:
> Starting some tests before doing the in-depth code review, below is an unordered list of findings:
> 
> The commands are:
> 
> 1. `export OCIO=/Users/hodoulp/dev/OpenColorIO-Configs/aces_1.0.3/config.ocio`
> 2. `lldb ./src/apps/ociodisplay/ociodisplay -- -v -gpuinfo -metal`
> 
> The findings are:
> 
> * using `ociodisplay` in debug mode and the [ACES 1.0.3](https://github.com/imageworks/OpenColorIO-Configs) config. The app successfully starts with some exceptions:
> 
> ```
> 2021-12-03 13:32:49.817127-0500 ociodisplay[13305:2504721] [General] -[__NSCFString replaceCharactersInRange:withString:]: Range or index out of bounds
> 2021-12-03 13:32:49.818170-0500 ociodisplay[13305:2504721] [General] (
> 	0   CoreFoundation                      0x00007fff375b4437 __exceptionPreprocess + 250
> 	1   libobjc.A.dylib                     0x00007fff7058c5bf objc_exception_throw + 48
> 	2   CoreFoundation                      0x00007fff37663e8e -[__NSCFString characterAtIndex:].cold.1 + 0
> 	3   CoreFoundation                      0x00007fff37664105 -[__NSCFString replaceOccurrencesOfString:withString:options:range:].cold.1 + 0
> 	4   CoreFoundation                      0x00007fff375f7178 mutateError + 44
> 	5   CoreFoundation                      0x00007fff3754acc8 -[__NSCFString replaceCharactersInRange:withString:] + 50
> 	6   GLUT                                0x00007fff39f76851 -[GLUTApplication applicationWillFinishLaunching:] + 175
> 	7   CoreFoundation                      0x00007fff3752e0af __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
> 	8   CoreFoundation                      0x00007fff3752e043 ___CFXRegistrationPost1_block_invoke + 63
> 	9   CoreFoundation                      0x00007fff3752dfb8 _CFXRegistrationPost1 + 372
> 	10  CoreFoundation                      0x00007fff3752dc24 ___CFXNotificationPost_block_invoke + 80
> 	11  CoreFoundation                      0x00007fff374fdd8d -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1554
> 	12  CoreFoundation                      0x00007fff374fd239 _CFXNotificationPost + 1351
> 	13  Foundation                          0x00007fff39b7d086 -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
> 	14  AppKit                              0x00007fff3479be99 -[NSApplication finishLaunching] + 330
> 	15  AppKit                              0x00007fff34a85911 _NSApplicationBeginRunning + 179
> 	16  AppKit                              0x00007fff34c3d125 -[NSApplication _beginRunning] + 26
> 	17  GLUT                                0x00007fff39f76523 -[GLUTApplication run] + 117
> 	18  GLUT                                0x00007fff39f827e7 glutMainLoop + 264
> 	19  ociodisplay                         0x0000000100007574 main + 2100
> 	20  libdyld.dylib                       0x00007fff71734cc9 start + 1
> )
> 2021-12-03 13:32:49.818638-0500 ociodisplay[13305:2504721] [General] _createMenuRef called with existing principal MenuRef already associated with menu
> 2021-12-03 13:32:49.819250-0500 ociodisplay[13305:2504721] [General] (
> 	0   CoreFoundation                      0x00007fff375b4437 __exceptionPreprocess + 250
> 	1   libobjc.A.dylib                     0x00007fff7058c5bf objc_exception_throw + 48
> 	2   CoreFoundation                      0x00007fff375b4295 +[NSException raise:format:] + 189
> 	3   AppKit                              0x00007fff347a3612 -[NSCarbonMenuImpl _createMenuRef] + 55
> 	4   AppKit                              0x00007fff347a2fe2 -[NSCarbonMenuImpl _instantiateCarbonMenu] + 133
> 	5   AppKit                              0x00007fff3479c005 -[NSApplication finishLaunching] + 694
> 	6   AppKit                              0x00007fff34a85911 _NSApplicationBeginRunning + 179
> 	7   AppKit                              0x00007fff34c3d125 -[NSApplication _beginRunning] + 26
> 	8   GLUT                                0x00007fff39f76523 -[GLUTApplication run] + 117
> 	9   GLUT                                0x00007fff39f827e7 glutMainLoop + 264
> 	10  ociodisplay                         0x0000000100007574 main + 2100
> 	11  libdyld.dylib                       0x00007fff71734cc9 start + 1
> )
> ```
> 
>  * right click on the grey scale image to change the transform to `DCDM`:
> 
> ```
> Process 13305 stopped
> * thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
>     frame #0: 0x00007fff7057f81d libobjc.A.dylib`objc_msgSend + 29
> libobjc.A.dylib`objc_msgSend:
> ->  0x7fff7057f81d <+29>: andl   0x18(%r10), %r11d
>     0x7fff7057f821 <+33>: shlq   $0x4, %r11
>     0x7fff7057f825 <+37>: addq   0x10(%r10), %r11
>     0x7fff7057f829 <+41>: cmpq   (%r11), %rsi
> Target 0: (ociodisplay) stopped.
> (lldb) thread backtrace
> * thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
>   * frame #0: 0x00007fff7057f81d libobjc.A.dylib`objc_msgSend + 29
>     frame #1: 0x00000001000226b1 ociodisplay`OpenColorIO_v2_2dev::MetalBuilder::applyColorCorrection(this=0x000000010913bc40, inputTexture=0x0000000119c21780, outputTexture=0x0000000119cf4570, outWidth=512, outHeight=512) at msl.mm:454:38
>     frame #2: 0x0000000100030d07 ociodisplay`OpenColorIO_v2_2dev::MetalApp::redisplay(this=0x0000000109025378) at metalapp.mm:366:25
>     frame #3: 0x00000001000052a2 ociodisplay`Redisplay() at main.cpp:209:19
>     frame #4: 0x00007fff39f6db21 GLUT`-[GLUTView drawRect:] + 93
>     frame #5: 0x00007fff39f6d88e GLUT`-[GLUTView handleWorkEvent:] + 3420
>     frame #6: 0x00007fff39f82592 GLUT`processWindowWorkList + 67
>     frame #7: 0x00007fff39f82530 GLUT`__glutProcessWorkEvents + 32
>     frame #8: 0x00007fff39f7658d GLUT`-[GLUTApplication run] + 223
>     frame #9: 0x00007fff39f827e7 GLUT`glutMainLoop + 264
>     frame #10: 0x0000000100007574 ociodisplay`main(argc=4, argv=0x00007ffeefbff938) at main.cpp:754:5
>     frame #11: 0x00007fff71734cc9 libdyld.dylib`start + 1
>     frame #12: 0x00007fff71734cc9 libdyld.dylib`start + 1
> (lldb) 
> ```
> 
> Last running `./tests/cpu/test_gpu_exce -metal` fully succeeds but it generates few identical messages:
> 
> * `[ 98/207] [GradingTone / style_log_highlights_rev            ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 0 at pix[0][0])`
> * `[148/207] [Lut1DOp / lut1d_1_small_legacy_shader             ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 6.34193e-05 at pix[41424][2]`
> * `[198/207] [MatrixOps / matrix_offset_generic_shader          ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 4.76837e-07 at pix[46153][1])`

@hodoulp  Thanks for all your comments. I applied them to the changelist. Let me know if there are other part of code that need to adopted.

May I ask if you are still getting `Context leak detected`? I am not able to reproduce this.

Morteeza commented on src/OpenColorIO/GpuShaderUtils.cpp:
```diff
@@ -435,6 +435,14 @@ std::string GpuShaderText::declareVarStr(const std::string & name, float v)
+{
+    std::string ret = lhs + " " + op + " " + rhs;
+    if(m_lang == GPU_LANGUAGE_MSL_2_0)
+        ret = "any( " + ret + " )";
```
one of the GPU test units produces following line in the MSL shader code and that causes compilation error in Metal.

```
float3 gamma = float3(0.790513813, 1.24223602, 0.828157365);
if ( gamma != float3(1., 1., 1.) )
{
```

To fix the compilation error, I added `any` that means if `any of the components are not equal`, then run the if-statement, otherwise jump over it or run the else-statement.

Morteeza commented:
> Starting some tests before doing the in-depth code review, below is an unordered list of findings:
> 
> The commands are:
> 
> 1. `export OCIO=/Users/hodoulp/dev/OpenColorIO-Configs/aces_1.0.3/config.ocio`
> 2. `lldb ./src/apps/ociodisplay/ociodisplay -- -v -gpuinfo -metal`
> 
> The findings are:
> 
> * using `ociodisplay` in debug mode and the [ACES 1.0.3](https://github.com/imageworks/OpenColorIO-Configs) config. The app successfully starts with some exceptions:
> 
> ```
> 2021-12-03 13:32:49.817127-0500 ociodisplay[13305:2504721] [General] -[__NSCFString replaceCharactersInRange:withString:]: Range or index out of bounds
> 2021-12-03 13:32:49.818170-0500 ociodisplay[13305:2504721] [General] (
> 	0   CoreFoundation                      0x00007fff375b4437 __exceptionPreprocess + 250
> 	1   libobjc.A.dylib                     0x00007fff7058c5bf objc_exception_throw + 48
> 	2   CoreFoundation                      0x00007fff37663e8e -[__NSCFString characterAtIndex:].cold.1 + 0
> 	3   CoreFoundation                      0x00007fff37664105 -[__NSCFString replaceOccurrencesOfString:withString:options:range:].cold.1 + 0
> 	4   CoreFoundation                      0x00007fff375f7178 mutateError + 44
> 	5   CoreFoundation                      0x00007fff3754acc8 -[__NSCFString replaceCharactersInRange:withString:] + 50
> 	6   GLUT                                0x00007fff39f76851 -[GLUTApplication applicationWillFinishLaunching:] + 175
> 	7   CoreFoundation                      0x00007fff3752e0af __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
> 	8   CoreFoundation                      0x00007fff3752e043 ___CFXRegistrationPost1_block_invoke + 63
> 	9   CoreFoundation                      0x00007fff3752dfb8 _CFXRegistrationPost1 + 372
> 	10  CoreFoundation                      0x00007fff3752dc24 ___CFXNotificationPost_block_invoke + 80
> 	11  CoreFoundation                      0x00007fff374fdd8d -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1554
> 	12  CoreFoundation                      0x00007fff374fd239 _CFXNotificationPost + 1351
> 	13  Foundation                          0x00007fff39b7d086 -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
> 	14  AppKit                              0x00007fff3479be99 -[NSApplication finishLaunching] + 330
> 	15  AppKit                              0x00007fff34a85911 _NSApplicationBeginRunning + 179
> 	16  AppKit                              0x00007fff34c3d125 -[NSApplication _beginRunning] + 26
> 	17  GLUT                                0x00007fff39f76523 -[GLUTApplication run] + 117
> 	18  GLUT                                0x00007fff39f827e7 glutMainLoop + 264
> 	19  ociodisplay                         0x0000000100007574 main + 2100
> 	20  libdyld.dylib                       0x00007fff71734cc9 start + 1
> )
> 2021-12-03 13:32:49.818638-0500 ociodisplay[13305:2504721] [General] _createMenuRef called with existing principal MenuRef already associated with menu
> 2021-12-03 13:32:49.819250-0500 ociodisplay[13305:2504721] [General] (
> 	0   CoreFoundation                      0x00007fff375b4437 __exceptionPreprocess + 250
> 	1   libobjc.A.dylib                     0x00007fff7058c5bf objc_exception_throw + 48
> 	2   CoreFoundation                      0x00007fff375b4295 +[NSException raise:format:] + 189
> 	3   AppKit                              0x00007fff347a3612 -[NSCarbonMenuImpl _createMenuRef] + 55
> 	4   AppKit                              0x00007fff347a2fe2 -[NSCarbonMenuImpl _instantiateCarbonMenu] + 133
> 	5   AppKit                              0x00007fff3479c005 -[NSApplication finishLaunching] + 694
> 	6   AppKit                              0x00007fff34a85911 _NSApplicationBeginRunning + 179
> 	7   AppKit                              0x00007fff34c3d125 -[NSApplication _beginRunning] + 26
> 	8   GLUT                                0x00007fff39f76523 -[GLUTApplication run] + 117
> 	9   GLUT                                0x00007fff39f827e7 glutMainLoop + 264
> 	10  ociodisplay                         0x0000000100007574 main + 2100
> 	11  libdyld.dylib                       0x00007fff71734cc9 start + 1
> )
> ```
> 
>  * right click on the grey scale image to change the transform to `DCDM`:
> 
> ```
> Process 13305 stopped
> * thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
>     frame #0: 0x00007fff7057f81d libobjc.A.dylib`objc_msgSend + 29
> libobjc.A.dylib`objc_msgSend:
> ->  0x7fff7057f81d <+29>: andl   0x18(%r10), %r11d
>     0x7fff7057f821 <+33>: shlq   $0x4, %r11
>     0x7fff7057f825 <+37>: addq   0x10(%r10), %r11
>     0x7fff7057f829 <+41>: cmpq   (%r11), %rsi
> Target 0: (ociodisplay) stopped.
> (lldb) thread backtrace
> * thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
>   * frame #0: 0x00007fff7057f81d libobjc.A.dylib`objc_msgSend + 29
>     frame #1: 0x00000001000226b1 ociodisplay`OpenColorIO_v2_2dev::MetalBuilder::applyColorCorrection(this=0x000000010913bc40, inputTexture=0x0000000119c21780, outputTexture=0x0000000119cf4570, outWidth=512, outHeight=512) at msl.mm:454:38
>     frame #2: 0x0000000100030d07 ociodisplay`OpenColorIO_v2_2dev::MetalApp::redisplay(this=0x0000000109025378) at metalapp.mm:366:25
>     frame #3: 0x00000001000052a2 ociodisplay`Redisplay() at main.cpp:209:19
>     frame #4: 0x00007fff39f6db21 GLUT`-[GLUTView drawRect:] + 93
>     frame #5: 0x00007fff39f6d88e GLUT`-[GLUTView handleWorkEvent:] + 3420
>     frame #6: 0x00007fff39f82592 GLUT`processWindowWorkList + 67
>     frame #7: 0x00007fff39f82530 GLUT`__glutProcessWorkEvents + 32
>     frame #8: 0x00007fff39f7658d GLUT`-[GLUTApplication run] + 223
>     frame #9: 0x00007fff39f827e7 GLUT`glutMainLoop + 264
>     frame #10: 0x0000000100007574 ociodisplay`main(argc=4, argv=0x00007ffeefbff938) at main.cpp:754:5
>     frame #11: 0x00007fff71734cc9 libdyld.dylib`start + 1
>     frame #12: 0x00007fff71734cc9 libdyld.dylib`start + 1
> (lldb) 
> ```
> 
> Last running `./tests/cpu/test_gpu_exce -metal` fully succeeds but it generates few identical messages:
> 
> * `[ 98/207] [GradingTone / style_log_highlights_rev            ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 0 at pix[0][0])`
> * `[148/207] [Lut1DOp / lut1d_1_small_legacy_shader             ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 6.34193e-05 at pix[41424][2]`
> * `[198/207] [MatrixOps / matrix_offset_generic_shader          ] - Context leak detected, msgtracer returned -1 PASSED - (MaxDiff: 4.76837e-07 at pix[46153][1])`

The exceptions you mentioned seems to be unrelated to this changelist. I am investigating why is this happening, but that shouldn't block this changelist.

hodoulp commented:
> May I ask if you are still getting Context leak detected? I am not able to reproduce this.
Yes. The same unit tests generate the exceptions.

hodoulp commented on src/libutils/oglapphelpers/metalapp.mm:
```diff
@@ -0,0 +1,391 @@
+    }
+    
+    setImageDimensions(imgWidth, imgHeight, comp);
+    m_image = std::make_shared<MtlTexture>(m_context->metalDevice, imgWidth, imgHeight, image);
```
[Question] The `image` pointer is dangling when leaving the method because it's the internal pointer to a local variable i.e. `imageData`. Does `MtlTexture` copy the data?

Morteeza replied:
internally `replaceRegion` will copy the image data immediately to the texture. There is no async operation happening here.

hodoulp commented on src/libutils/oglapphelpers/msl.mm:
```diff
@@ -0,0 +1,504 @@
+    }
+    
+    return float4AdaptedLutValues;
+}
```
[Comment] The memory management of the image data introduces at least one useless copy. As `float4AdaptedLutValues` is a local variable the return could make a copy. With C++11 one could use `std::move` -- or -- the new buffer could be an in/out parameter of the method!

hodoulp commented on src/libutils/oglapphelpers/msl.mm:
```diff
@@ -0,0 +1,504 @@
+    [tex replaceRegion:MTLRegionMake3D(0, 0, 0, edgelen, edgelen, edgelen)
+           mipmapLevel:0
+                 slice:0
+             withBytes:texData
```
[Question] As `texData` is the internal pointer from `float4AdaptedLutValues` which is a local variable, it will be a dangling pointer when leaving the method scope. Does MTLTexture copy the image data?

Morteeza replied:
replaceRegion will copy the image data immediately to the texture. There is no async operation happening here.

hodoulp commented on src/libutils/oglapphelpers/mtltexture.h:
```diff
@@ -0,0 +1,75 @@
+    
+    int         getWidth()       const { return m_width;  }
+    int         getHeight()      const { return m_height; }
+    uint64_t    getGLHandle()    const { assert(m_openGLContext); return m_texID; }
```
Please replace the assert by an exception.

hodoulp commented:
[Question] Input and Output images are using `MtlTexture` but LUTs are using `MTLTexture`. Why? Is there a behavior difference?

hodoulp commented on src/libutils/oglapphelpers/metalapp.mm:
```diff
@@ -0,0 +1,391 @@
+        std::ostringstream uniformParams;
+        std::ostringstream params;
+        
+        main << "//Metal Shading Language version 2.3\n"
```
The MSL translation is based on MSL 2.0 but the comment indicates MSL 2.3. The generated shader program should also be 2.0 to be consistent. Is there a need to be 2.3?

Morteeza replied:
Wow!!! Good Catch :D

hodoulp commented on src/libutils/oglapphelpers/msl.mm:
```diff
@@ -0,0 +1,504 @@
+    [[renderPipelineDesc colorAttachments][0] setPixelFormat:MTLPixelFormatRGBA32Float];
+    m_PSO = [[m_device newRenderPipelineStateWithDescriptor:renderPipelineDesc error:&error] autorelease];
+    
+    return (error == nil);
```
The code should throw an exception to better trap the issue; otherwise, `MetalApp::redisplay()` will still try to apply a color correction even after the failure.

hodoulp commented on src/libutils/oglapphelpers/metalapp.mm:
```diff
@@ -0,0 +1,391 @@
+                                            m_outputImage->getHeight());
+    }
+    
+    if(!m_glStateBound)
```
:bangbang: `buildPipelineStateObject` sets the variable `m_glStateBound` to true; but there is no code setting it back to false when a new color transformation is requested. So, the texture bindings are never done for the new color transformation!

Morteeza replied:
`m_glStateBound` is independent of Metal calls and state. This only initialises the shader that takes the output of metal calls and show it in GLUT window. The shader doesn't change with switching the color transformation. The only construct that OpenGL state is dependent on is the metal output target textureId in OpenGL. As far as I can see, this object live span is same as the whole application. So, changing `m_glStateBound` after it being initialised is not really needed.

hodoulp commented on src/libutils/oglapphelpers/metalapp.h:
```diff
@@ -0,0 +1,73 @@
+
+private:
+    MetalBuilderRcPtr m_metalBuilder;
+    bool m_glStateBound;   // OpenGL state for outputing the metal output texture contents is bound
```
:bangbang: The default value is missing. It should be false by default.

hodoulp commented on src/libutils/oglapphelpers/metalapp.mm:
```diff
@@ -0,0 +1,391 @@
+    
+    if(!m_glStateBound)
+    {
+        prepareAndBindOpenGLState();
```
Should the bindings be done before the `applyColorCorrection()` ?

Morteeza replied:
It actually doesn't matter. This sets OpenGL state and acts independent from applyColorCorrection.

Morteeza commented on src/libutils/oglapphelpers/msl.mm:
```diff
@@ -0,0 +1,504 @@
+    samplerStateCache[interpolation] = [[device newSamplerStateWithDescriptor:samplerDesc] autorelease];
+    
+    return samplerStateCache[interpolation];
+}
```
replaceRegion will copy the image data immediately to the texture. There is no async operation happening here.

hodoulp commented on tests/gpu/GPUUnitTest.cpp:
```diff
@@ -228,8 +228,13 @@ namespace
             // It means to generate the input values.
 
             const bool testWideRange = test->getTestWideRange();
+#if __aarch64__
```
The define is for ARM processors only. Should that be `__aarch64__ && __APPLE__` ?

Morteeza replied:
yeah. Better to limit it to Apple.

hodoulp reviewed (approved):
Great job @Morteeza

remia reviewed (commented):
Hi @Morteeza,

Sorry for the late review, great work! Here is some minor points on the code.

I tried it locally on macOS 12.0.1 and it appears to work fine. I just noticed the following message while running GPU unit tests: `Context leak detected, CoreAnalytics returned false`, it happen for `matrix_inverse`, `lin2log`, `style_log_midtones_rev` but do not affect the result (tests successful). Do you see the same thing?

michdolan commented on src/libutils/oglapphelpers/oglapp.h:
```diff
@@ -87,16 +87,16 @@ class OglApp
     // Initialize the image.
-    void initImage(int imageWidth, int imageHeight,
+    virtual void initImage(int imageWidth, int imageHeight,
                    Components comp, const float * imageBuffer);
```
```suggestion
                           Components comp, const float * imageBuffer);
```

michdolan replied:
(super minor, so could be fixed later...)

michdolan reviewed (approved):
Thanks for your contribution @Morteeza !

Morteeza commented on src/libutils/oglapphelpers/msl.mm:
```diff
@@ -0,0 +1,534 @@
+id<MTLSamplerState> GetSamplerState(id<MTLDevice> device, Interpolation interpolation)
+{
+    MTLSamplerDescriptor* samplerDesc = [MTLSamplerDescriptor new];
+    [samplerDesc setMinFilter:MTLSamplerMinMagFilterNearest];
```
Good point! :) Thanks 👍

Morteeza commented on src/libutils/oglapphelpers/msl.mm:
```diff
@@ -0,0 +1,534 @@
+
+void MetalBuilder::deleteAllTextures()
+{
+    m_textureIds.clear();
```
That's right. Fixed.

Morteeza commented on src/libutils/oglapphelpers/metalapp.mm:
```diff
@@ -0,0 +1,399 @@
+    
+    
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_RECTANGLE_EXT, m_outputImage->getGLHandle());
```
This texture is shared between gl and metal. The way metal works, we create pixel buffer and then take an OpenGL handle, and a Metal texture handle from it. The way metal demands us to use this texture in OpenGL is to use its as a rectangle texture. It can't be used as a normal texture2d.

Morteeza commented on src/libutils/oglapphelpers/metalapp.mm:
```diff
@@ -0,0 +1,399 @@
+        std::ostringstream uniformParams;
+        std::ostringstream params;
+        
+        main << "//Metal Shading Language version 2.0\n"
```
it does. Fixed it.

remia commented:
Thanks for the quick fixes @Morteeza! The Context leak warning can be addressed later if problematic.

Morteeza commented:
> Thanks for the quick fixes @Morteeza! The Context leak warning can be addressed later if problematic.

Thanks for your comments. I will investigate that more. Unfortunately, I can't reproduce it on my machine. I have to see why.
